Cassandra datatype handler

taherkl opened the pull request:
The Type handler will be used for Casting source type(Spanner) to Cassandra Cassandra-supported type.

This code covers the support for:

- Boolean
- Float
- Double
- Date
- String
- Text
- Timestamp
- Integer
- Varchar, 
- ascii, 
- inet, 
- smallint, 
- tinyint
- List of String, Date, Text, Integer, Date, Timestamp, Float, Double
- Set of String, Date, Text, Integer, Date, Timestamp, Float, Double

// TODO - We are going to create a common method which will be the common integration point of this class and then we will mark all of these methods as private. we have kept this method public for a while to wire up UT.

VardhanThigle commented on v2/spanner-to-sourcedb/src/main/java/com/google/cloud/teleport/v2/templates/dbutils/dml/CassandraTypeHandler.java:
```diff
@@ -0,0 +1,466 @@
+     * @return a {@link Date} object containing the value represented in cassandra type.
+     */
+    public static Date handleCassandraDateType(String colName, JSONObject valuesJson) {
+        return handleCassandraGenericDateType(colName, valuesJson, "yyyy-MM-dd");
```
I might be missing the complete context here.

In case this is going to be passed to Cassadra 4.0 Driver later on,
Would it make sense to [Local Date](https://docs.datastax.com/en/drivers/java/4.1/com/datastax/oss/driver/api/core/data/SettableById.html#setLocalDate-com.datastax.oss.driver.api.core.CqlIdentifier-java.time.LocalDate-)

For Some of the temporal types, we might need to be careful of the differences in how time zones are handled in Spanner and Cassandra.

For example, Cassandra Date is free of time zone.

pawankashyapollion replied:
Yes @VardhanThigle we are using the LocalDate here

VardhanThigle replied:
Thanks. Looks like this got addressed as a part of [pull/2054/commits/5f9affa3566411e196cafe420fb55fc918d20a0a](https://github.com/GoogleCloudPlatform/DataflowTemplates/pull/2054/commits/5f9affa3566411e196cafe420fb55fc918d20a0a). Resolving.

VardhanThigle commented on v2/spanner-to-sourcedb/src/main/java/com/google/cloud/teleport/v2/templates/dbutils/dml/CassandraTypeHandler.java:
```diff
@@ -0,0 +1,466 @@
+     *
+     * @return a {@link Date} object containing timestamp as value represented in cassandra type.
+     */
+    public static Date handleCassandraTimestampType(String colName, JSONObject valuesJson) {
```
Similar to https://github.com/GoogleCloudPlatform/DataflowTemplates/pull/2054/files#r1877268541,
Would `Instant` work better?

VardhanThigle replied:
The `handleCassandraTimestampType` is now returning `LocalDate`.
As per the [driver docs](https://javadoc.io/static/com.datastax.oss/java-driver-core/4.13.0/com/datastax/oss/driver/api/core/data/SettableById.html) `setInstant` works by default for TimeStamp type.

pawankashyapollion replied:
@VardhanThigle  we are doing this while assigning this value to cassandra DML Statement

VardhanThigle replied:
Would you be creating the DML as a query string or would you use the driver to set the values into a statement?
I would strongly lean on setting the values as it helps avoid un-intended sql injections and overall helps us make best use of Cassandra driver's type safety.

I don't have complete context on how we might do this when we actually write to Cassandra, to help ease our conversation, here's an example of a possible write - You would definitely have more elaborate ways that supply you the column names and values than the example here. 
## Example
### Table
Imagine we have a table like:
 ```sql
 CREATE TABLE IF NOT EXISTS example.data_types
                    id int PRIMARY KEY, 
                    time_col time, 
                    timestamp_col timestamp,
                    date_col date,
                    integer_col int,
                    medium_int_col int, 
                    tiny_int_col tinyint; 
```
### Java Example
```java 
        // Create an immutable list of column names
        List<String> columnNames = ImmutableList.of("id", "time_col", "timestamp_col", "date_col", "integer_col", "medium_int_col", "tiny_int_col");

        // Join the column names with commas
        String columns = String.join(", ", columnNames);           // Columns
        
      String questionMarks = columnNames.stream()
                .map(name -> "?")
                .collect(Collectors.joining(", "));
            // Prepare the statement
            PreparedStatement prepared = session.prepare("INSERT INTO example.data_types (" + columns ") " +
                    "VALUES (" + questionMarks + ")");

            // Bind values and execute
            // Here the values need to be strongly typed, for example for `Date` we need `LocalDate`.
            BoundStatement bound = prepared.bind(
                    1,
                    LocalTime.now(),
                    Instant.now(),
                    LocalDate.now(),
                    123456789,
                    (short) 25000,  // Medium int value (using Short)
                    (byte) 120      // Tiny int value (using Byte)
            );
```

## Something we might want to avoid
We might want to avoid creating a query that directly fills in the values into the string, bypassing the drivers settable interface (which helps with safety)

pawankashyapollion replied:
@VardhanThigle we are doing the same way , I guess it will be cleared in DML PR

taherkl replied:
Thank you for sharing this detailed explanation @VardhanThigle . Our implementation is aligned with your suggestions which will be shared for review in our upcoming PRs. Appreciate it.

VardhanThigle replied:
Thanks for the confirmation on prepared statements @taherkl .
In case we go with prepared statements, we will need to map incoming records from spanner to the types supported by the Cassandra 4.0 driver API.

On the read path I have experienced that the driver throws codec mismatch exceptions even for a mismatch like reading a `medium int` as an `int` instead of a `short`. Likewise on the Date, Time and related types too.

That was the motivation behind few of the review comments to map `TimeStamp` to java `Instant` and the like.

In case you are planning to reconvert it to a compatible type before calling the prepared statement in a future PR, it might be better to get to a compatible class in this class itself (in the scope of current PR to the extent possible)

If we do type marshallings at multiple places, possible customer issues around loss of precision (for time related types) or overflow/undeflow etc get very hard to debug. To the extent possible let's keep the conversion from spanner to Cassandra type at limited places.

taherkl replied:
Yes, @VardhanThigle The return types will be bonded directly to the prepared statements and there won't be another layer of conversion. We will take care of that. We want to short PRs and hence you may have many questions about it. Please review the future PRs, which shall clear most of the confusions and address your suggestions.

VardhanThigle replied:
Thanks for the confirmation.
I can see that for timestamp we are still returning `LocalDate`
 [ref](https://github.com/GoogleCloudPlatform/DataflowTemplates/pull/2054/commits/957b8866afe58d464a0ace86dde6e1062337a4e7#diff-690e435057a9227e15eccda34696e95bebf8b51afe6a0ac421227a392276dce1R261)
 
 There's the Cassandra driver doc that I have linked to in most of these discussions around types. It's [here](https://javadoc.io/static/com.datastax.oss/java-driver-core/4.13.0/com/datastax/oss/driver/api/core/data/SettableById.html)
 
 As per the driver interface [setInstant](https://javadoc.io/static/com.datastax.oss/java-driver-core/4.13.0/com/datastax/oss/driver/api/core/data/SettableById.html#setInstant-com.datastax.oss.driver.api.core.CqlIdentifier-java.time.Instant-) method is supported for the CQL timestamp type
 
 where as,
 [setLocalDate](https://javadoc.io/static/com.datastax.oss/java-driver-core/4.13.0/com/datastax/oss/driver/api/core/data/SettableById.html#setLocalDate-com.datastax.oss.driver.api.core.CqlIdentifier-java.time.LocalDate-) method is supported for CQL date type.
 
 In case we use the `LocalDate` type and call `setLocalDate` for timestamp type on CQL, I am afraid, the driver might throw a codec mismatch exception. Same applies to mediumInt (works with Java `Short`) and tinyint (works with Java `byte`) type.
 
 
 The request here is to map the types to the ones expected by the Cassandra 4.0 driver.

pawankashyapollion replied:
@VardhanThigle as we metioned earlier we are handling some unboxing stuff while assigning the object to DML
For example: we are parsing it to integer and while assiging based on sourceType we are casting to smallint , tinyint etc. we are doing the same in terms of timestamp as well

shreyakhajanchi replied:
What is the advantage of converting it to LocalDate and then Instant in this case , why not use Instant directly 

<img width="712" alt="image" src="https://github.com/user-attachments/assets/d2517cee-cccb-4c9c-be50-b24dee73719f" />

pawankashyapollion replied:
@shreyakhajanchi, we aimed to generalize the approach by converting date and time or timestamp to a LocalDate, which can later be converted as needed for convenience but as of I now I have address as per suggestion

VardhanThigle replied:
Thanks.  This has been addressed now.
Let's directly map spanner objects to the types that are expected by Cassandra Driver. This will reduce any debugging for type handling or precisions at a single place.

I have had a best effort look at the changes in this PR but please check that all the types are compliant with the driver specification before merging.

VardhanThigle commented on v2/spanner-to-sourcedb/src/main/java/com/google/cloud/teleport/v2/templates/dbutils/dml/CassandraTypeHandler.java:
```diff
@@ -0,0 +1,466 @@
+        return handleCassandraGenericDateType(colName, valuesJson, "yyyy-MM-dd'T'HH:mm:ss.SSSZ");
+    }
+
+    private static Date handleCassandraGenericDateType(String colName, JSONObject valuesJson, String formatter) {
```
This common translation Function seems to be servicing 2 different Cassandra Types,  from 2 possible input Json Types. 
It might help to add a bit of javadoc on what spanner types map to a formatted date and what map to long, and how we convert either of them to Date (for specific Cassandra types)

Unless absolutely necessary let's try to avoid this kind of coupling. In case it helps, please add more comments that help explain the context.

pawankashyapollion replied:
@VardhanThigle we have added more detailed comments

VardhanThigle commented on v2/spanner-to-sourcedb/src/main/java/com/google/cloud/teleport/v2/templates/dbutils/dml/CassandraTypeHandler.java:
```diff
@@ -0,0 +1,466 @@
+     * @return a {@link Integer} object containing Integer as value represented in cassandra type.
+     */
+    public static Integer handleCassandraIntType(String colName, JSONObject valuesJson) {
+        return valuesJson.getBigInteger(colName).intValue();
```
In case we are using Cassandra driver to write, atleast on the read path, I have experienced it doing strong type checks for `tinyInt` and `MediumInt` with java `Byte` and `Short`.
On Spanner side all of the smaller integer version will map to Integer alone.
It might make sense to add separate handling for such types, and in case of possibility of truncation, log error/ Sever DLQ entries.

pawankashyapollion replied:
hi @VardhanThigle We are converting the value coming from Spanner to Integer and while assigning we are converting to Byte and Short under Typehandler mapping getColumnValueByType
<img width="655" alt="Screenshot 2024-12-10 at 11 47 25 AM" src="https://github.com/user-attachments/assets/6c125d93-563f-4a34-b201-c6d7b87b49f6">

VardhanThigle replied:
Thanks. Is this snippet from a future PR?
Would it make sense to put all the conversions at one place?
As in what if we try to convert it to `short` or `byte` in this class itself?
That way all the conversions needed to convert from Spanner to Cassandra compatible types will be in a single place (to the extent possible).

pawankashyapollion replied:
@VardhanThigle  we are doing in the same class  which will be provide more visibility in further PR

pawankashyapollion replied:
@shreyakhajanchi We have added the rest datatypes as well in the same PR

VardhanThigle commented on v2/spanner-to-sourcedb/src/main/java/com/google/cloud/teleport/v2/templates/dbutils/dml/CassandraTypeHandler.java:
```diff
@@ -0,0 +1,466 @@
+     *
+     * @return a {@link List<Long>} object containing List<Long> as value represented in cassandra type.
+     */
+    public static List<Long> handleInt64ArrayType(String colName, JSONObject valuesJson) {
```
Would it make sense to use Generics to handle array types?

For example, in case there's a generic of kind.
`handleArrayType<T> (String columnName, JSONObject valueJson)`
`handleSetType<T>` becomes an extension of the same.

There are various ways to achieve it. One possibility is to have an interface, which has a default implementation for handling sets, based on the overridden implementation for arrays.

pawankashyapollion replied:
Addressed the suggestion @VardhanThigle

VardhanThigle replied:
Could you please point me to the commit or change where this is handled?

VardhanThigle replied:
Thanks.

VardhanThigle commented on v2/spanner-to-sourcedb/src/main/java/com/google/cloud/teleport/v2/templates/dbutils/dml/CassandraTypeHandler.java:
```diff
@@ -0,0 +1,466 @@
+    public static Boolean handleCassandraBoolType(String colName, JSONObject valuesJson) {
+        return valuesJson.getBoolean(colName);
+    }
+
```
There are few more types like `Inet` and `Duration`  etc.
In case we are handling them in a future PR, please add a TODO for the same.

taherkl replied:
Hi @VardhanThigle . Yes, there will be a future PR with the rest of the datatypes. Will add a TODO for rest of the types.

pawankashyapollion replied:
@VardhanThigle we have add the missing types as you mentioned

shreyakhajanchi commented on v2/spanner-to-sourcedb/src/main/java/com/google/cloud/teleport/v2/templates/dbutils/dml/CassandraTypeHandler.java:
```diff
@@ -0,0 +1,515 @@
+     * @return a {@link Double} object containing the value represented in cassandra type.
+     */
+    public static Double handleCassandraDoubleType(String colName, JSONObject valuesJson) {
+        return valuesJson.getBigDecimal(colName).doubleValue();
```
In case the column value is null valuesJson.getBigDecimal() will throw an exception right ? Shouldn't we use  something like `optBigDecimal` here

pawankashyapollion replied:
we have handle the null case fo all the case @shreyakhajanchi

taherkl replied:
Hi @shreyakhajanchi Resolving this as we have addressed the comment, Thanks

shreyakhajanchi commented on v2/spanner-to-sourcedb/src/test/java/com/google/cloud/teleport/v2/templates/dbutils/dml/CassandraTypeHandlerTest.java:
```diff
@@ -0,0 +1,478 @@
+import static org.junit.Assert.*;
+
+@RunWith(JUnit4.class)
+public class CassandraTypeHandlerTest {
```
Add more tests covering cases for NULL values for all datatypes, min and max values, incorrect value formats, etc. Please refer to these tests for forward migration for an idea on test cases to cover: https://github.com/GoogleCloudPlatform/DataflowTemplates/blob/main/v2/spanner-common/src/test/java/com/google/cloud/teleport/v2/spanner/migrations/convertors/ChangeEventTypeConvertorTest.java

pawankashyapollion replied:
handled the same @shreyakhajanchi

shreyakhajanchi commented on v2/spanner-to-sourcedb/src/test/java/com/google/cloud/teleport/v2/templates/dbutils/dml/CassandraTypeHandlerTest.java:
```diff
@@ -0,0 +1,478 @@
+        assertEquals(expected, result);
+    }
+
+    @Test
+    public void testByteArrayHandling_ValidByteArray() {
+        byte[] inputByteArray = {1, 2, 3, 4, 5};
+        Object colValue = inputByteArray;
+        byte[] byteArray = null;
+        if (colValue instanceof byte[]) {
+            byteArray = (byte[]) colValue;
+        }
+        assertNotNull(byteArray);
+        assertArrayEquals(inputByteArray, byteArray);
+    }
+
+    @Test
+    public void testByteArrayHandling_NotByteArray() {
+        String inputValue = "This is a string";
+        Object colValue = inputValue;
+        byte[] byteArray = null;
+        if (colValue instanceof byte[]) {
+            byteArray = (byte[]) colValue;
+        }
+        assertNull(byteArray);
+    }
+
```
Which method is being unit tested here? Same comment for few other methods

pawankashyapollion replied:
Address the same @shreyakhajanchi

shreyakhajanchi commented on v2/spanner-to-sourcedb/src/test/java/com/google/cloud/teleport/v2/templates/dbutils/dml/CassandraTypeHandlerTest.java:
```diff
@@ -0,0 +1,478 @@
+    public void testHandleFloatSetType_InvalidString() {
+        String newValuesString = "{\"values\":[\"1.1\", \"2.2\", \"abc\"]}";
+        JSONObject newValuesJson = new JSONObject(newValuesString);
+        try {
+            handleFloatSetType("values", newValuesJson);
+            fail("Expected IllegalArgumentException for invalid number format");
+        } catch (IllegalArgumentException e) {
+            assertTrue(e.getMessage().contains("Invalid number format for column values"));
```
Instead of wrapping these in try catch block , use something like:
```
@Rule public ExpectedException expectedEx = ExpectedException.none();
expectedEx.expect(IllegalArgumentException.class);
expectedEx.expectMessage("Invalid number format for column values");
``` 
Same comment for all methods throwing exception

pawankashyapollion replied:
Address the same @shreyakhajanchi

pawankashyapollion commented on v2/spanner-to-sourcedb/src/main/java/com/google/cloud/teleport/v2/templates/dbutils/dml/CassandraTypeHandler.java:
```diff
@@ -0,0 +1,515 @@
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package com.google.cloud.teleport.v2.templates.dbutils.dml;
```
Sure @shreyakhajanchi

pawankashyapollion replied:
Address the same @shreyakhajanchi

shreyakhajanchi replied:
The github actions are still failing

pawankashyapollion replied:
@shreyakhajanchi We ran build locally and its passed. can we retry now

pawankashyapollion commented on v2/spanner-to-sourcedb/src/main/java/com/google/cloud/teleport/v2/templates/dbutils/dml/CassandraTypeHandler.java:
```diff
@@ -0,0 +1,515 @@
+import java.time.LocalDate;
+import java.time.format.DateTimeFormatter;
+import java.time.format.DateTimeParseException;
+import java.util.*;
```
Sure @shreyakhajanchi

pawankashyapollion replied:
Address the same @shreyakhajanchi

pawankashyapollion commented on v2/spanner-to-sourcedb/src/main/java/com/google/cloud/teleport/v2/templates/dbutils/dml/CassandraTypeHandler.java:
```diff
@@ -0,0 +1,515 @@
+     * @return a {@link List<Boolean>} object containing List<Boolean> as value represented in cassandra type.
+     */
+    public static List<Boolean> handleBoolArrayType(String colName, JSONObject valuesJson) {
+        return valuesJson.getJSONArray(colName).toList().stream()
```
Address the same @shreyakhajanchi

taherkl commented:
@shreyakhajanchi @VardhanThigle We have addressed all the review comments. Requesting your review for the same. Please close the comment if the addressed changes are aligned with the expectations. Thanks!

shreyakhajanchi commented on v2/spanner-to-sourcedb/src/main/java/com/google/cloud/teleport/v2/templates/dbutils/dml/CassandraTypeHandler.java:
```diff
@@ -0,0 +1,940 @@
+        colName,
+        valuesJson,
+        obj -> {
+          if (obj instanceof Number) {
```
add a case for obj instanceof Long

pawankashyapollion replied:
Address @shreyakhajanchi

shreyakhajanchi commented on v2/spanner-to-sourcedb/src/test/java/com/google/cloud/teleport/v2/templates/dbutils/dml/CassandraTypeHandlerTest.java:
```diff
@@ -0,0 +1,981 @@
+    handleCassandraInetAddressType(colKey, newValuesJson);
+  }
+
+  @Test(expected = IllegalArgumentException.class)
```
Add a happy test case for handleStringifiedJsonToMap

shreyakhajanchi commented on v2/spanner-to-sourcedb/src/main/java/com/google/cloud/teleport/v2/templates/dbutils/dml/CassandraTypeHandler.java:
```diff
@@ -0,0 +1,940 @@
+   * @return A string representation of the timestamp in UTC that is compatible with Cassandra.
+   * @throws RuntimeException If the timestamp string is invalid or the conversion fails.
+   */
+  private static String convertToCassandraTimestamp(String value, String timezoneOffset) {
```
This method is set to be private and I see no usage for it in the class, same for other method with the same name

shreyakhajanchi replied:
Same comment for multiple other private methods

shreyakhajanchi replied:
I think you need to revisit the access modifiers, method like `handleCassandraGenericDateType` should be `public` and utility methods like `convertToCassandraTimestamp` should be `private`

pawankashyapollion replied:
Yes @shreyakhajanchi as I mentioned already we are revisiting the access modifiers in DML PR as we have one method which is going to handle all case so rest all will be private except one mapping method

shreyakhajanchi commented on v2/spanner-to-sourcedb/src/test/java/com/google/cloud/teleport/v2/templates/dbutils/dml/CassandraTypeHandlerTest.java:
```diff
@@ -0,0 +1,981 @@
+  @Test
+  public void testConvertToCassandraTimestampWithValidOffset() {
+    String value = "2024-12-12T10:15:30+02:00";
+    String timezoneOffset = "+00:00";
```
Test with a non-zero offset

pawankashyapollion replied:
@shreyakhajanchi in cassandra we do not have offset I guess

shreyakhajanchi replied:
In that case why do you have this method:
`public static String convertToCassandraTimestamp(String value, String timezoneOffset) `

pawankashyapollion replied:
Agree @shreyakhajanchi we are removing extra method in DML PR which is not required as of now we have added all posibility

shreyakhajanchi commented on v2/spanner-to-sourcedb/src/main/java/com/google/cloud/teleport/v2/templates/dbutils/dml/CassandraTypeHandler.java:
```diff
@@ -0,0 +1,940 @@
+   * @param value The string to be escaped.
+   * @return The escaped string where single quotes are replaced with double single quotes.
+   */
+  private static String escapeCassandraString(String value) {
```
This is a private method where is it being used?

pawankashyapollion replied:
this is being used in mapping part

shreyakhajanchi commented on v2/spanner-to-sourcedb/src/main/java/com/google/cloud/teleport/v2/templates/dbutils/dml/CassandraTypeHandler.java:
```diff
@@ -0,0 +1,940 @@
+   */
+  public static Float handleCassandraFloatType(String colName, JSONObject valuesJson) {
+    try {
+      return valuesJson.getBigDecimal(colName).floatValue();
+    } catch (JSONException e) {
+      return null;
+    }
+  }
+
+  /**
+   * Generates a Type based on the provided {@link CassandraTypeHandler}.
+   *
+   * @param colName - which is used to fetch Key from valueJSON.
+   * @param valuesJson - contains all the key value for current incoming stream.
+   * @return a {@link Double} object containing the value represented in cassandra type.
+   */
+  public static Double handleCassandraDoubleType(String colName, JSONObject valuesJson) {
+    try {
+      return valuesJson.getBigDecimal(colName).doubleValue();
+    } catch (JSONException e) {
+      return null;
+    }
```
Why are we returning in case of exception rather than using optBigDecimal ?

pawankashyapollion replied:
optBigDecimal => return Null in case of value is null while parsing so I have added this to handle null pointer execption

VardhanThigle commented on v2/spanner-to-sourcedb/src/main/java/com/google/cloud/teleport/v2/templates/dbutils/dml/CassandraTypeHandler.java:
```diff
@@ -0,0 +1,943 @@
+   */
+  public static InetAddress handleCassandraInetAddressType(String colName, JSONObject valuesJson)
+      throws UnknownHostException {
+    return InetAddress.getByName(valuesJson.getString(colName));
```
Please consider using [com.google.common.net.InetAddresses#forString](https://guava.dev/releases/20.0/api/docs/com/google/common/net/InetAddresses.html#forString-java.lang.String-] instead of `InetAddress.getByName`

Motivation:
The [InetAddress.html#getByName(java.lang.String)](https://docs.oracle.com/javase/7/docs/api/java/net/InetAddress.html#getByName(java.lang.String)) That's used here tries to resolve a host name into it's IP address via a DNS query. 

Most likely customer's Dataflow workers will not have firewall permissions to do do that.  `getByName` will cause the code to throw [UnknownHostException](https://docs.oracle.com/javase/7/docs/api/java/net/UnknownHostException.html) which we are not catching here.
Depending on the configuration of the security manager on customer's jvm, `getByName` can also throw  [SecurityException](https://docs.oracle.com/javase/7/docs/api/java/lang/SecurityException.html) both of which we would like to avoid having to deal with.

The `google.common` equivalent [com.google.common.net.InetAddresses#forString](https://guava.dev/releases/20.0/api/docs/com/google/common/net/InetAddresses.html#forString-java.lang.String-]  does not do any DNS query and simply throws [IllegalArgumentException](http://docs.oracle.com/javase/7/docs/api/java/lang/IllegalArgumentException.html?is-external=true) if the string isn't a valid ipaddress literal.

VardhanThigle replied:
Thanks. This has been address now.

VardhanThigle commented on v2/spanner-to-sourcedb/src/main/java/com/google/cloud/teleport/v2/templates/dbutils/dml/CassandraTypeHandler.java:
```diff
@@ -0,0 +1,943 @@
+   * @return {@code true} if the string is a valid IP address, {@code false} otherwise.
+   */
+  public static boolean isValidIPAddress(String value) {
+    try {
```
This involves a DNS query, let's avoid this by using
[com.google.common.net.InetAddresses#forString(lava.lang.String)](https://guava.dev/releases/20.0/api/docs/com/google/common/net/InetAddresses.html#forString-java.lang.String-)

pawankashyapollion replied:
Address the same @VardhanThigle

VardhanThigle replied:
Thanks. As this has migrated to the `com.google.common.net.InetAddresses#forString` it's fine with me.
I am assuming in the larger picture you are calling this to validate. Resolving.

VardhanThigle commented on v2/spanner-to-sourcedb/src/test/java/com/google/cloud/teleport/v2/templates/dbutils/dml/CassandraTypeHandlerTest.java:
```diff
@@ -0,0 +1,1028 @@
+    assertEquals(expected, handleCassandraInetAddressType(colKey, newValuesJson));
+  }
+
+  @Test(expected = UnknownHostException.class)
```
in case you go with [com.google.common.net.InetAddresses](https://guava.dev/releases/20.0/api/docs/com/google/common/net/InetAddresses.html#forString-java.lang.String-), this will change to [IllegalArgumentException](http://docs.oracle.com/javase/7/docs/api/java/lang/IllegalArgumentException.html?is-external=true)

Also Consider adding a case for `www.google.com`. In case we go with the  [com.google.common.net.InetAddresses](https://guava.dev/releases/20.0/api/docs/com/google/common/net/InetAddresses.html#forString-java.lang.String-), this will also throw `IllegalArgumentException` since the `google.common` package avoids DNS lookup

@shreyakhajanchi this would mean, that we support only ip addresses for inet columns and not host names 
I would treat this as a fair tradeoff given that Cassandra only supports  Inet that `Represents an IP address, either IPv4 (4 bytes long) or IPv6 (16 bytes long) `
with but we need to document that

VardhanThigle replied:
Please consider adding a test to check that "www.google.com" is treated as an invalid IP address to ensure that the underlying library we use does not attempt any DNS queries.

VardhanThigle commented on v2/spanner-to-sourcedb/src/test/java/com/google/cloud/teleport/v2/templates/dbutils/dml/CassandraTypeHandlerTest.java:
```diff
@@ -0,0 +1,1318 @@
+  public void testHandleCassandraInetAddressType_Hostname() {
+    JSONObject valuesJson = new JSONObject();
+    valuesJson.put("col1", "www.google.com");
+    IllegalArgumentException exception =
```
Thanks for adding this test case.

[super-nit]
Could we consistenty use `assertThrows` for the unit tests which need asserts?

`assertThrows` is the strongly preferred idiom. [go/java-testing/exceptions#assertThrows](http://go/java-testing/exceptions#assertThrows) And
[exceptions#discouraged](http://go/java-testing/exceptions#discouraged)

taherkl replied:
Sure we will make this consistent change in the following PRs

shreyakhajanchi commented on v2/spanner-to-sourcedb/src/main/java/com/google/cloud/teleport/v2/templates/dbutils/dml/CassandraTypeHandler.java:
```diff
@@ -0,0 +1,974 @@
+   * @param valuesJson - contains all the key value for current incoming stream.
+   * @return a {@link Long} object containing Long as value represented in cassandra type.
+   */
+  public static Long handleCassandraBigintType(String colName, JSONObject valuesJson) {
```
It should be of type BigInteger: https://docs.datastax.com/en/drivers/java/4.1/com/datastax/oss/driver/api/core/data/SettableByName.html#setBigInteger-java.lang.String-java.math.BigInteger-

pawankashyapollion replied:
@shreyakhajanchi  As per the design document, we need to convert int64 values to bigint.
In Cassandra, the bigint type corresponds to a 64-bit signed integer, which can be handled in Java using the Long type.
Therefore, I have used Long to represent the int64 values, as Long is the Java equivalent for handling 64-bit integer values.
<meta charset="utf-8"><b style="font-weight:normal;" id="docs-internal-guid-66f8c448-7fff-00fd-3956-ce24a3329bd4"><h4 dir="ltr" style="line-height:1.38;background-color:#ffffff;margin-top:0pt;margin-bottom:4pt;padding:10pt 0pt 0pt 0pt;"></h4><div dir="ltr" style="margin-left:0pt;" align="left">
Spanner Type | Cassandra Type
-- | --
Int64 | Bigint

</div></b>

I guess it will become more clear with DML PR

VardhanThigle replied:
For this particular type, `setLong` works too
[SettableByName.html#setLong-java.lang.String-long-](https://docs.datastax.com/en/drivers/java/4.1/com/datastax/oss/driver/api/core/data/SettableByName.html#setLong-java.lang.String-long-)

shreyakhajanchi commented on v2/spanner-to-sourcedb/src/main/java/com/google/cloud/teleport/v2/templates/dbutils/dml/CassandraTypeHandler.java:
```diff
@@ -0,0 +1,974 @@
+import org.json.JSONException;
+import org.json.JSONObject;
+
+class CassandraTypeHandler {
```
make the class public static to avoid importing each methods individually and just import the class

pawankashyapollion replied:
@shreyakhajanchi Yes we have done this in DML PR as I mentioned earlier, that, all of the methods become private except the one method which will be entrypoint for all the type

shreyakhajanchi commented on v2/spanner-to-sourcedb/src/main/java/com/google/cloud/teleport/v2/templates/dbutils/dml/CassandraTypeHandler.java:
```diff
@@ -0,0 +1,974 @@
+   * @param colName - which is used to fetch Key from valueJSON.
+   * @param valuesJson - contains all the key value for current incoming stream.
+   * @return a {@link Long} object containing Long as value represented in cassandra type.
+   */
```
Missing handling for some cassandra types like long, BigDecimal, etc. Please ensure that all types from this doc are covered: https://docs.datastax.com/en/drivers/java/4.1/com/datastax/oss/driver/api/core/data/SettableByName.html

pawankashyapollion replied:
we are handling the missing case in our DML PR
